codec/decode.go

```diff
@@ -2413,8 +2413,24 @@ func (d *Decoder) rawBytes() []byte {
 	return bs2
 }
 
-func (d *Decoder) wrapErrstr(v interface{}, err *error) {
-	*err = fmt.Errorf("%s decode error [pos %d]: %v", d.hh.Name(), d.r.numread(), v)
+// DecodeError wraps any error encountered during decoding.
+type DecodeError struct {
+	Name string
+	Pos  int
+	Err  error
+}
+
```

Comment:
Is there a really good reason for exporting this type and its fields? If not, let's keep the type and its fields unexported, and just expose the Cause() method if it helps.

Reply:
if this isn't exported, then there's no way to get the name and pos without string parsing the error, and this PR is all about avoiding having to string-parse

Reply:
Currently I have no need for pulling the name/pos, though -- all I need is the Cause() method.